hbase-server/src/main/java/org/apache/hadoop/hbase/security/provider/DigestSaslServerAuthenticationProvider.java

```diff
@@ -0,0 +1,155 @@
+  private static final Logger LOG = LoggerFactory.getLogger(
+      DigestSaslServerAuthenticationProvider.class);
+
+  private AtomicReference<UserGroupInformation> attemptingUser = new AtomicReference<>(null);
+
+  @Override
+  public AttemptingUserProvidingSaslServer createServer(
+      SecretManager<TokenIdentifier> secretManager,
+      Map<String, String> saslProps) throws IOException {
+    if (secretManager == null) {
+      throw new AccessDeniedException("Server is not configured to do DIGEST authentication.");
+    }
+    final SaslServer server = Sasl.createSaslServer(getSaslAuthMethod().getSaslMechanism(), null,
+      SaslUtil.SASL_DEFAULT_REALM, saslProps,
+      new SaslDigestCallbackHandler(secretManager, attemptingUser));
+
+    return new AttemptingUserProvidingSaslServer(server, () -> attemptingUser.get());
+  }
+
+  /** CallbackHandler for SASL DIGEST-MD5 mechanism */
+  private static class SaslDigestCallbackHandler implements CallbackHandler {
+    private final SecretManager<TokenIdentifier> secretManager;
+    private final AtomicReference<UserGroupInformation> attemptingUser;
+
+    public SaslDigestCallbackHandler(SecretManager<TokenIdentifier> secretManager,
+        AtomicReference<UserGroupInformation> attemptingUser) {
+      this.secretManager = secretManager;
+      this.attemptingUser = attemptingUser;
+    }
+
+    private char[] getPassword(TokenIdentifier tokenid) throws InvalidToken {
+      return SaslUtil.encodePassword(secretManager.retrievePassword(tokenid));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void handle(Callback[] callbacks) throws InvalidToken, UnsupportedCallbackException {
+      NameCallback nc = null;
+      PasswordCallback pc = null;
+      AuthorizeCallback ac = null;
+      for (Callback callback : callbacks) {
+        if (callback instanceof AuthorizeCallback) {
+          ac = (AuthorizeCallback) callback;
+        } else if (callback instanceof NameCallback) {
+          nc = (NameCallback) callback;
+        } else if (callback instanceof PasswordCallback) {
+          pc = (PasswordCallback) callback;
+        } else if (callback instanceof RealmCallback) {
+          continue; // realm is ignored
+        } else {
+          throw new UnsupportedCallbackException(callback, "Unrecognized SASL DIGEST-MD5 Callback");
+        }
+      }
+      if (pc != null) {
+        TokenIdentifier tokenIdentifier = HBaseSaslRpcServer.getIdentifier(
+            nc.getDefaultName(), secretManager);
+        attemptingUser.set(tokenIdentifier.getUser());
+        char[] password = getPassword(tokenIdentifier);
+        if (LOG.isTraceEnabled()) {
+          LOG.trace("SASL server DIGEST-MD5 callback: setting password " + "for client: " +
+              tokenIdentifier.getUser());
+        }
+        pc.setPassword(password);
+      }
+      if (ac != null) {
+        String authid = ac.getAuthenticationID();
+        String authzid = ac.getAuthorizationID();
```

Comment:
For readability, can these two parameter `authid` and `authzid` be named more distinguishable

Reply:
Cleaned up this method and also added some comments about what it's actually doing.